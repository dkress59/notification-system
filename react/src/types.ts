import { DetailedHTMLProps, HTMLAttributes, PropsWithChildren } from 'react'

import { SpawnBannerProps, SpawnModalProps, SpawnToastProps } from ':core/types'

type CustomHTMLAttributes<
	T,
	K = void | Record<string, unknown>,
> = DetailedHTMLProps<HTMLAttributes<T> & K, T>

export type BannerAreaAttributes = CustomHTMLAttributes<HTMLBannerAreaElement>

export type NotificationAreaAttributes =
	CustomHTMLAttributes<HTMLNotificationAreaElement>

export type BannerNotificationAttributes =
	CustomHTMLAttributes<HTMLBannerNotificationElement>

export type ModalNotificationAttributes =
	CustomHTMLAttributes<HTMLModalNotificationElement>

export type ToastNotificationAttributes =
	CustomHTMLAttributes<HTMLToastNotificationElement>

declare global {
	// eslint-disable-next-line @typescript-eslint/no-namespace
	namespace JSX {
		interface IntrinsicElements {
			['banner-area']: BannerAreaAttributes
			['notification-area']: NotificationAreaAttributes
			['banner-notification']: BannerNotificationAttributes
			['modal-notification']: ModalNotificationAttributes
			['toast-notification']: ToastNotificationAttributes
		}
	}
}

export type BannerAreaProps = PropsWithChildren

export type NotificationAreaProps = PropsWithChildren

export interface BannerNotificationProps
	extends PropsWithChildren,
		Omit<SpawnBannerProps, 'content'> {
	onDismiss?: () => void
}

export interface ToastNotificationProps
	extends PropsWithChildren,
		Omit<SpawnToastProps, 'content'> {
	onDismiss?: () => void
}

export interface ModalNotificationProps
	extends PropsWithChildren,
		Omit<SpawnModalProps, 'content'> {
	onConfirm?: () => void
	onDecline?: () => void
	onDismiss?: () => void
}
