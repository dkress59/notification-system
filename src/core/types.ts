import {
	HTMLBannerAreaElement,
	HTMLBannerNotificationElement,
	HTMLModalNotificationElement,
	HTMLNotificationAreaElement,
	HTMLToastNotificationElement,
} from './components'

type SnakeToCamelCaseProp<S extends string> = S extends
	| `${infer T} ${infer U}`
	| `${infer T}-${infer U}`
	? `${Uncapitalize<T>}${Capitalize<SnakeToCamelCaseProp<U>>}`
	: Uncapitalize<S>
type SnakeToCamelCase<T> = {
	[K in keyof T as SnakeToCamelCaseProp<K & string>]: T[K]
}

export enum NotificationType {
	SUCCESS = 'success',
	INFO = 'info',
	ERROR = 'error',
	WARNING = 'warning',
}

export enum NotificationEvent {
	BANNER_DISMISSED = 'bannerDismissed',
	MODAL_CONFIRMED = 'modalConfirmed',
	MODAL_DECLINED = 'modalDeclined',
	MODAL_DISMISSED = 'modalDismissed',
	TOAST_DISMISSED = 'toastDismissed',
}

type SpawnArgs<T> = T & {
	content: string
}

export interface BannerAttributes {
	'auto-hide'?: boolean | string
	'auto-hide-after-ms'?: number | string
	'headline'?: string
	'type'?: NotificationType
}
export type SpawnBannerArgs = SpawnArgs<SnakeToCamelCase<BannerAttributes>>

export interface ModalAttributes {
	'condition'?: boolean
	'headline'?: string
	'show-confirm'?: boolean
	'show-decline'?: boolean
	'type'?: NotificationType
	'label-confirm'?: string
	'label-decline'?: string
}
export type SpawnModalArgs = SpawnArgs<SnakeToCamelCase<ModalAttributes>>

export interface ToastAttributes {
	'auto-hide'?: boolean | string
	'auto-hide-after-ms'?: number | string
	'headline'?: string
	'type'?: NotificationType
}
export type SpawnToastArgs = SpawnArgs<SnakeToCamelCase<ToastAttributes>>

declare global {
	interface HTMLElementTagNameMap {
		'banner-area': HTMLBannerAreaElement
		'banner-notification': HTMLBannerNotificationElement
		'modal-notification': HTMLModalNotificationElement
		'notification-area': HTMLNotificationAreaElement
		'toast-notification': HTMLToastNotificationElement
	}
}
