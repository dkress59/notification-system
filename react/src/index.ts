export * from './components/banner-area'
export * from './components/banner-notification'
export * from './components/modal-notification'
export * from './components/notification-area'
export * from './components/toast-notification'
export * from './types'
