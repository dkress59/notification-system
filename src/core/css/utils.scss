%text {
  &-grey {
    color: var(--clr-grey);
  }

  &-green {
    color: var(--clr-green);

    &-light {
      color: var(--clr-green-light);
    }
  }

  &-blue {
    color: var(--clr-blue);

    &-light {
      color: var(--clr-blue-light);
    }
  }

  &-red {
    color: var(--clr-red);

    &-light {
      color: var(--clr-red-light);
    }
  }

  &-yellow {
    color: var(--clr-yellow);

    &-light {
      color: var(--clr-yellow-light);
    }
  }
}

%bg {
  &-grey {
    background-color: var(--clr-grey);
  }

  &-green {
    background-color: var(--clr-green);

    &-light {
      background-color: var(--clr-green-light);
    }
  }

  &-blue {
    background-color: var(--clr-blue);

    &-light {
      background-color: var(--clr-blue-light);
    }
  }

  &-red {
    background-color: var(--clr-red);

    &-light {
      background-color: var(--clr-red-light);
    }
  }

  &-yellow {
    background-color: var(--clr-yellow);

    &-light {
      background-color: var(--clr-yellow-light);
    }
  }
}

%border {
  &-grey {
    border-color: var(--clr-grey);
  }

  &-green {
    border-color: var(--clr-green);

    &-light {
      border-color: var(--clr-green-light);
    }
  }

  &-blue {
    border-color: var(--clr-blue);

    &-light {
      border-color: var(--clr-blue-light);
    }
  }

  &-red {
    border-color: var(--clr-red);

    &-light {
      border-color: var(--clr-red-light);
    }
  }

  &-yellow {
    border-color: var(--clr-yellow);

    &-light {
      border-color: var(--clr-yellow-light);
    }
  }
}

%min-w-60 {
  min-width: 240px;
}

%shadow-lg {
  box-shadow: 0 10px 15px -3px rgb(0 0 0 / 10%), 0 4px 6px -4px rgb(0 0 0 / 10%);
}

%opacity-0 {
  opacity: 0;
}

%opacity-100 {
  opacity: 100;
}
