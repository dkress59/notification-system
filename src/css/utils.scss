@tailwind base;
@tailwind components;
@tailwind utilities;

@layer utilities {
  .custom {
    &-toast {
      @apply relative right-0 w-max max-w-1/2 min-w-60 my-4 mx-8;

      &:first-of-type {
        @apply mt-8;
      }
    }

    &-modal {
      @apply fixed top-1/2 left-1/2 w-1/2 h-1/2 min-w-60 max-w-2xl;
      @apply transform-gpu -translate-x-1/2 -translate-y-1/2;
    }

    &-toast,
    &-modal {
      @apply py-4 px-12 pointer-events-auto border-2 rounded-md shadow-lg;
      @apply transition-all duration-500;

      &-type {
        &-success {
          @apply bg-green-light border-green;

          .icon {
            @apply text-green border-green;
          }
        }

        &-info {
          @apply bg-blue-light border-blue;

          .icon {
            @apply text-blue border-blue;
          }
        }

        &-error {
          @apply bg-red-light border-red;

          .icon {
            @apply text-red border-red;
          }
        }

        &-warning {
          @apply bg-yellow-light border-yellow;

          .icon {
            @apply text-yellow border-yellow;
          }
        }
      }

      >.icon {
        @apply absolute top-4 left-4 h-6 aspect-square rounded-full overflow-hidden border-2;
        @apply text-center text-xs leading-5 font-bold;
      }

      >h4 {
        @apply font-bold text-lg text-grey leading-snug;
      }

      >div {
        @apply text-grey;
      }

      >.dismiss {
        @apply absolute top-4 right-4 font-bold cursor-pointer;
      }
    }
  }
}
