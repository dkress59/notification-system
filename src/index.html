<!DOCTYPE html>
<html dir="ltr" lang="en">

<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=5.0" />
  <title>Notification Web Components</title>

  <script type="module" src="/build/notification-system.esm.js"></script>
  <script nomodule src="/build/notification-system.js"></script>

  <style>
    * {
      box-sizing: border-box;
    }

    body {
      font-family: sans-serif;
    }

    #notification-container {
      position: fixed;
      z-index: 50;
      top: 0;
      right: 0;
      bottom: 0;
      left: 0;
      pointer-events: none;
      display: flex;
      flex-direction: column;
      align-items: flex-end;
      overflow-y: auto;
    }

    /* #notification-container > * {
        right: 0;
      } */
  </style>
  <script>
    document.addEventListener('DOMContentLoaded', () => {
      document.querySelectorAll('notify-toast').forEach(toast => {
        toast.addEventListener('toastDismissed', (shadowRoot) => console.debug(
          'toast dismissed!',
          { shadowRoot }
        ))
      })
    })
  </script>
</head>

<body>
  <div id="notification-container">
    <!-- <my-component first="Stencil" last="'Don't call me a framework' JS"></my-component>
      <my-component first="more Stencil" middle="please" last="'Don't call me a framework' JS"></my-component> -->
    <notify-toast type="error" headline="Lorem Ipsum">Lorem ipsum dolor sit amet consectetur adipisicing elit, facilis
      necessitatibus sunt quisquam officia quidem.</notify-toast>
    <notify-toast>Mock Content</notify-toast>
    <notify-toast auto-hide>Mock Content</notify-toast>
    <notify-toast type="info" is-hidden>Hidden Content</notify-toast>
    <notify-toast>Mock Content</notify-toast>
  </div>
</body>

</html>