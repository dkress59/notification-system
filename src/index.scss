@import 'css/base.scss';
// @import 'css/util.scss';
