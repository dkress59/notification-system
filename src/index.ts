export { Components, JSX } from './components'
export { BannerArea } from './components/banner-area/banner-area'
export { BannerNotification } from './components/banner-notification/banner-notification'
export { ModalNotification } from './components/modal-notification/modal-notification'
export { NotificationArea } from './components/notification-area/notification-area'
export { ToastNotification } from './components/toast-notification/toast-notification'
