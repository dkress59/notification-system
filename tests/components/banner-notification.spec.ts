import { fireEvent, screen } from '@testing-library/dom'

import { NotificationEvent, NotificationType } from '../../src/types'
import { getIconElement } from '../../src/utils'

describe('<banner-notification />', () => {
	beforeEach(() => {
		document.body.innerHTML = ''
	})
	it('shadow root matches snapshot', () => {
		const notification = document.createElement('banner-notification')
		notification.setAttribute('data-testid', 'notification')
		notification.setAttribute('headline', 'mock_headline')
		notification.setAttribute('type', 'error')
		notification.innerHTML = '<p>slotted_paragraph</p>'
		document.body.appendChild(notification)
		const element = screen.getByTestId('notification')
		const dismissButton = element.shadowRoot!.querySelector('.dismiss')
		expect(dismissButton).not.toBeNull()
		expect(element.shadowRoot?.innerHTML).toMatchSnapshot()
	})
	it('fades in', () => {
		jest.useFakeTimers()
		const notification = document.createElement('banner-notification')
		notification.setAttribute('data-testid', 'notification')
		document.body.appendChild(notification)
		const element = screen.getByTestId('notification')
		expect(element.classList).toContain('hidden')
		jest.runOnlyPendingTimers()
		expect(element.classList).not.toContain('hidden')
		jest.useRealTimers()
	})
	it('has changeable type', () => {
		const initialType = NotificationType.INFO
		const changedType = NotificationType.ERROR
		const notification = document.createElement('banner-notification')
		notification.setAttribute('data-testid', 'notification')
		notification.setAttribute('type', initialType)
		document.body.appendChild(notification)
		expect(
			screen
				.getByTestId('notification')
				.shadowRoot!.querySelector('.icon')!,
		).toEqual(getIconElement(initialType))

		notification.setAttribute('type', changedType)
		expect(
			screen
				.getByTestId('notification')
				.shadowRoot!.querySelector('.icon')!,
		).toEqual(getIconElement(changedType))
	})
	it('has changeable headline', () => {
		const initialHeadline = 'mock_headline'
		const changedHeadline = 'mock_headline_changed'
		const notification = document.createElement('banner-notification')
		notification.setAttribute('data-testid', 'notification')
		notification.setAttribute('headline', initialHeadline)
		document.body.appendChild(notification)
		expect(
			screen.getByTestId('notification').shadowRoot!.querySelector('h4'),
		).toHaveTextContent(initialHeadline)

		notification.setAttribute('headline', changedHeadline)
		expect(
			screen.getByTestId('notification').shadowRoot!.querySelector('h4'),
		).toHaveTextContent(changedHeadline)
	})
	it('hides dismiss button if auto-hide is set', () => {
		const notification = document.createElement('banner-notification')
		notification.setAttribute('data-testid', 'notification')
		notification.setAttribute('auto-hide', '')
		document.body.appendChild(notification)
		const element = screen.getByTestId('notification')
		const dismissButton = element.shadowRoot!.querySelector('.dismiss')
		expect(dismissButton).toBeNull()
	})
	it('auto-hides', () => {
		jest.useFakeTimers()
		const notification = document.createElement('banner-notification')
		notification.setAttribute('data-testid', 'notification')
		notification.setAttribute('auto-hide', '')
		document.body.appendChild(notification)
		jest.runAllTimers()
		fireEvent.transitionEnd(notification)
		const element = screen.queryByTestId('notification')
		expect(element).toBeNull()
		jest.useRealTimers()
	})
	it('can be dismissed programmatically', () => {
		jest.useFakeTimers()
		const notification = document.createElement('banner-notification')
		notification.setAttribute('data-testid', 'notification')
		document.body.appendChild(notification)
		const element = screen.getByTestId('notification')
		expect(element.classList).toContain('hidden')
		jest.runOnlyPendingTimers()
		expect(element.classList).not.toContain('hidden')
		notification.dismiss()
		expect(element.classList).toContain('hidden')
		jest.useRealTimers()
	})
	it('fires "bannerDismissed" event', () => {
		const mockCallback = jest.fn()
		const notification = document.createElement('banner-notification')
		notification.setAttribute('data-testid', 'notification')
		document.body.appendChild(notification)
		const element = screen.getByTestId('notification')
		element.addEventListener(
			NotificationEvent.BANNER_DISMISSED,
			mockCallback,
		)
		expect(mockCallback).not.toHaveBeenCalled()
		notification.dismiss()
		fireEvent.transitionEnd(notification)
		expect(mockCallback).toHaveBeenCalled()
	})
	it('removes itself from the DOM after dismissal', () => {
		jest.useFakeTimers()
		const notification = document.createElement('banner-notification')
		notification.setAttribute('data-testid', 'notification')
		document.body.appendChild(notification)
		jest.runAllTimers()
		const element = screen.getByTestId('notification')
		const dismissButton = element.shadowRoot!.querySelector('.dismiss')!
		fireEvent.mouseDown(dismissButton)
		fireEvent.transitionEnd(notification)
		expect(screen.queryByTestId('notification')).toBeNull()
		jest.useRealTimers()
	})
})
