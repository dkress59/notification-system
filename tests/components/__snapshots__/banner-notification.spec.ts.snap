// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<banner-notification /> shadow root matches snapshot 1`] = `
"<style>mock-style {
	mock-prop: value;
}</style><h4 data-testid=\\"headline\\">mock_headline</h4><span data-testid=\\"icon-type\\" class=\\"icon\\"><span class=\\"error\\">╲</span></span><section><slot></slot></section><span class=\\"dismiss\\" data-testid=\\"btn-dismiss\\" tabindex=\\"0\\" aria-role=\\"button\\">✕</span>"
`;
